### PR TITLE
on Firefox choose screen sharing mode

### PR DESCRIPTION
**Pull request by hrotkogabor**
On Firefox, the user should choose what to share: full screen, or a single window.
Like here:
https://janus.conf.meetecho.com/screensharingtest.html

This pull request needs to be work together with the jitsi-meet's firefox_share_fullscreen branch.

### CODE REVIEW

**Comment by hrotkogabor** [comment]
I have sent the signed CLA.

**Comment by hrotkogabor** [comment]
I have sent the signed CLA once again.

**Comment by damencho** [comment]
Jenkins: add to whitelist

File: JitsiMeetJS.js [hunk 1]
```diff
@@ -385,6 +385,7 @@ var LibJitsiMeet = {
     util: {
         ScriptUtil: ScriptUtil,
         RTCUIHelper: RTCUIHelper,
+        RTCBrowserType: RTCBrowserType,
```
Comment (damencho) [inline JitsiMeetJS.js]: Exposing RTCBrowserType from the lib is a long running internal discussion and we do not want to expose it for now. So can you use a workaround in the jitsi-meet code, where this thing is used and do not expose RTCBrowserType?
There is similar thing here https://github.com/jitsi/jitsi-meet/blob/8745efb81f7aa4fb1790610a47b083722a871645/modules/UI/videolayout/SmallVideo.js#L648-L648
It is not a clean solution, but expanding the API of the lib and then removing stuff from the API can be more painful than fixing those hacks.

**Comment by paweldomas** [comment]
Could you please also rebase the PR to get rid of merge commits ?

**Comment by tuxayo** [comment]
@paweldomas How would you proceed to get rid of the merge commits? (I fail understand Solution 1 here: http://stackoverflow.com/a/24393326/3682839)

**Comment by paweldomas** [comment]
I would rebase hrotkogabor:firefox_share_fullscreen on top of jitsi:master and then force push, but I don't know the exact commands at this time. I can rebase it and merge myself if that's too much trouble for you.

**Comment by saghul** [comment]
You can do an interactive rebase and squash all your commits into one by running `git rebase -i origin/master` in your branch and marking all the commits with an `s`.  Then edit the commit message from right there and once done force push.

**Comment by tuxayo** [comment]
> I would rebase hrotkogabor:firefox_share_fullscreen on top of jitsi:master and then force push

That wouldn't get rid of the merge commits, would it?

> You can do an interactive rebase and squash all your commits into one

Wouldn't that pollute the diff of the remaining commit? hrotkogabor changes would be mixed with the merge. I'm not sure, that's the limit of my git knowledge.

> I can rebase it and merge myself if that's too much trouble for you.

I don't have the time now but if @hrotkogabor don't have the time either I'll try to do it and submit the result to them so they can update this PR.

**Comment by martinszy** [comment]
Hello, I was just trying to share the full screen in firefox and I can't, can someone please merge this?

**Comment by tuxayo** [comment]
For anyone not interested in this:
If you have few tens of minutes you can try to fork hrotkogabor's repo and rebase/merge this branch with jisti:master.

I did this a few months ago without any knowledge about the project and that allowed to keep this patch on track. But it needs to be done again.

So with basic git and conflict resolution knowledge, anyone can push this patch forward. And if you're unconfident/beginner, that's a great exercise to learn while contributing to a libre/open source project.
And if you are not certain of some choices in the conflict resolution, just write them down and post then in a message here.

**Comment by hrotkogabor** [comment]
The basic idea of this patch should work, but I do not have the time to fix and test it sorry.

**Comment by tuxayo** [comment]
Can you or anyone else with the knowledge can the steps to manually test that it works? i.e. is there more to do besides deploying locally jitsi-meet and trying a full screensharing between 2 firefox browsers?

Are there other cases to test for this?

**Comment by saghul** [comment]
> Hello, I was just trying to share the full screen in firefox and I can't, can someone please merge this?

We cannot merge it as-is, because it has merge conflicts. Either the original athor fixes them and updates this PR, or someone else submits a new PR, with his / her permission, of course.

**Comment by tuxayo** [comment]
Hi here is a branch with a merge of master in this PR:
https://github.com/tuxayo/lib-jitsi-meet/commits/firefox_share_fullscreen

### Conflict resolution
#### in obtainAudioAndVideoPermissions(options = {})
When calling obtainDevices()
I changed options to 'options: options'
But I'm not sure because upstream changed 'deviceGUM: deviceGUM' to 'deviceGUM'
So I don't know if it affects options

#### in obtainScreenOnFirefox()
I'm not sure options should be an argument as this.option also exists.
But I don't know if it's the same thing.

#### obtainWebRTCScreen(streamCallback, failCallback, options)
option has been kept last param (to follow @hrotkogabor's work)
A check has been done (only in lib-jitsi-meet) to ensure that there are no calls to this function with different args.

#### obtainScreenOnFirefox(options, callback, errorCallback, options)
option has been kept last param (to follow @hrotkogabor's work)
A check has been done (only in lib-jitsi-meet) to ensure that there are no calls to this function with different args.

#### gumFunction()
Now it takes options as last param
This function was called GUM before
in onGetStreamResponse() there is a call to gumFunction and it also take a last options argument. So it seems consistent.

As check was also done for any call to non existing GUM() function that the merge could have reintroduced.

#### @ anyone
if the conflict resolutions are unclear and call into question the reliability of the merge. I'm ready to redo it following another method:
- merge and commit the conflicts
- solve them in a separate commit to make them reviewable.
I wish I though about this before.

### @hrotkogabor 
Can you review my work and merge `tuxayo/firefox_share_fullscreen` into your branch to update this PR? (it's going to be a fast forward merge so no merge commit)
Or is there another way to move forward?
Including submitting my own PR.
Or ticking the checkbox to allow the Jitsi maintainer to write to your branch and merge mine into it.

### @damencho 

Is this still relevant/up to date? https://github.com/jitsi/lib-jitsi-meet/pull/129#pullrequestreview-18292442

**Comment by hrotkogabor** [comment]
I added  tuxayo as a collaborator to my repo.

**Comment by saghul** [comment]
Thanks! So, @tuxayo can you please make a new PR or rebase this one?

**Comment by tuxayo** [comment]
@hrotkogabor Thanks, I interpret that has being able to push my merge into this PR's branch. Is that correct? (I don't want to add mess because of a misunderstanding)

**Comment by tuxayo** [comment]
@hrotkogabor I think you meant to add me as a collaborator to the `lib-jitsi-meet` repo but you did on the `jitsi-meet` one.

**Comment by hrotkogabor** [comment]
I added you to lib-jitsi-meet also.

**Comment by tuxayo** [comment]
Thanks, I've updated this PR.

@saghul or any other maintainer: What's next?

**Comment by saghul** [comment]
Jenkins please test.

**Comment by tuxayo** [comment]
Maybe Jenkins in on holiday :-P

**Comment by bbaldino** [comment]
Jenkins, test this please

**Comment by damencho** [comment]
There are few lint errors:

06:21:07 /home/jenkins/lib-jitsi-meet/modules/RTC/RTCUtils.js
06:21:07    296:49  error  Unexpected trailing comma    comma-dangle
06:21:07   1093:21  error  Expected property shorthand  object-shorthand
06:21:07 
06:21:07 /home/jenkins/lib-jitsi-meet/modules/RTC/ScreenObtainer.js
06:21:07   282:1  error  Line 282 exceeds the maximum line length of 80  max-len
06:21:07 
06:21:07 ✖ 3 problems (3 errors, 0 warnings)

**Comment by damencho** [comment]
I tried testing this and I got an error, is it working for you?:

[JitsiMeetJS.js] <getGlobalOnErrorHandler>:  UnhandledError: TypeError: e is not a function "Script: https://my/libs/lib-jitsi-meet.min.js?v=139" Line: 2 Column: 315802 StackTrace:  TypeError: e is not a function
Stack trace:
r/<@https://my/libs/lib-jitsi-meet.min.js?v=139:2:315802
value/<@https://my/libs/lib-jitsi-meet.min.js?v=139:2:66763
h@https://my/libs/lib-jitsi-meet.min.js?v=139:2:59469
m/</<@https://my/libs/lib-jitsi-meet.min.js?v=139:2:59702

Seems to me the error is inside ScreenObtiner.obtainWebRTCScreen

**Comment by tuxayo** [comment]
> I tried testing this and I got an error, is it working for you?:

How did you try ? I'm stuck with #573

> Seems to me the error is inside ScreenObtiner.obtainWebRTCScreen

How do you manage to see that with the minified version?

I checked the call to obtainWebRTCScreen() and they are consistent.
The caller is obtainScreenOnFirefox() but I don't see what is it's caller. (jitsi-meet doesn't seem to call it directly either)

**Comment by tuxayo** [comment]
After running the tests, I got an error: https://framagit.org/snippets/648

But I have the same on master: https://framagit.org/snippets/647

**Comment by damencho** [comment]
I had deployed it on my development deployment and tried sharing desktop with Firefox.

**Comment by tuxayo** [comment]
Isn't there a dependency with this patch's counterpart? (which is behind upstream with conflicts to be solved as I (wrongly?) hoped to work on both sequentially)

As stated in this initial message
https://github.com/jitsi/lib-jitsi-meet/pull/129#issue-159363720
> This pull request needs to be work together with the jitsi-meet's firefox_share_fullscreen branch.

@hrotkogabor can you confirm whether or not the two patch must be merged at the same time in lib-jitsi-meet and jitsi-meet to work?
Or is the lib-jitsi-meet's part supposed to be backward compatible with jitsi-meet's?

**Comment by johnxie** [comment]
Is screen share supported for jitsi meet on Firefox?

**Comment by saghul** [comment]
Closing since this is incredibly outdated. We do support screensharing on Firefox.

**Comment by johnxie** [comment]
Thanks, noted, the video-sharing button is not working in the latest
Firefox 79.0 (64-bit). Will dig deeper.

**Pull request closed without merging**